.github/workflows/ci-pipeline.yml

```diff
@@ -7,10 +7,29 @@ on:
     paths-ignore: 'README.md'
 
 jobs:
-  build:
+  build-and-test:
     name: Build and test
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v4
     - run: docker build --target test --tag todo-app:test .
-    - run: docker run todo-app:test
\ No newline at end of file
+    - run: docker run todo-app:test
+  
+  push-and-deploy:
+    name: Push to Docker Hub and deploy
+    runs-on: ubuntu-latest
+    needs: build-and-test
+    if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main'}}
+    steps: 
+    - uses: actions/checkout@v4
+    - name: Login to Docker Hub
+      uses: docker/login-action@v3
+      with:
+        username: ${{ vars.DOCKERHUB_USERNAME }}
```

Comment:
Nice use of a variable, most people use a secret here and while there's no harm it always seems a little odd when the username is usually quite easy to work out anyway!